Homework1

panda1987DS opened the pull request:
За основы был взят пример "ml_project_example" 
Приложение исполняется из консоли.
Структура и примеры исполнения подробно описаны в файле README.

**Код разбит на следующие части:**
- чтение параметров работы
- чтение, разделение данных
- подготовка фичей (делается с помощью трансформеров, добавлен вручную написанный трансформер SquaringTransformer, который возводит фичу в квадрат)
- обучение модели, предсказание, вычисление метрик, сохранение модели на диск
- пайплайн, который вызывает все фунции по очереди, а также логирует процесс работы.

Все логи пишутся в файл train_model.log, а также в стандартный вывод. Сообщения выше worning также логируются в файл train_model.warm

**Предусмотрено три режима работы:**
1. Полный:
- считываются данные
- делятся на обучение и тест
- обрабатываются фичи
- обучается модель
- обученная модель сохраняется на диск
- выполняются предсказания
- считаются метрики на тестовой части данных, а также кросс- валидация на 10 вариантах разбития данных
- все метрики сохраняются в логи и файл метрик 

2. train: 
- считываются данные
- делятся на обучение и тест
- обрабатываются фичи
- обучается модель
- обученная модель сохраняется на диск

3. predict:

- считываются данные
- делятся на обучение и тест
- обрабатываются фичи
- считывается ранее сохраненная модель
- выполняются предсказания
- считаются метрики на тестовой части данных, а также кросс- валидация на 10 вариантах разбития данных
- все метрики сохраняются в логи и файл метрик
 

**_Самооценка:_**

-2) Назовите ветку homework1   1 балл **+**
-1) положите код в папку ml_project
0) В описании к пулл реквесту описаны основные "архитектурные" и тактические решения, которые сделаны в вашей работе. В общем, описание что именно вы сделали и для чего, чтобы вашим ревьюерам было легче понять ваш код. (2 балла) **+**

1) Выполнение EDA, закоммитьте ноутбук в папку с ноутбуками (2 баллов)
Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы)
Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + 1 балл)
- _**не до конца поняла данный пункт, можно ли считать файлы с метриками отчетами. Возможно 0 баллов**_  
2) Проект имеет модульную структуру(не все в одном файле =) ) (2 баллов) **+**

3) использованы логгеры (2 балла) **+**

4) написаны тесты на отдельные модули и на прогон всего пайплайна(3 баллов) **+**

5) Для тестов генерируются синтетические данные, приближенные к реальным (3 баллов)
- можно посмотреть на библиотеки https://faker.readthedocs.io/en/, https://feature-forge.readthedocs.io/en/latest/
- можно просто руками посоздавать данных, собственноручно написанными функциями
как альтернатива, можно закоммитить файл с подмножеством трейна(это не оценивается) 
**+, правда сторонние библиотеки не использовала, т.к. сами данные довольно простые**

6) Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing) (3 балла)
**+**

7) Используются датаклассы для сущностей из конфига, а не голые dict (3 балла) **+**

8) Используйте кастомный трансформер(написанный своими руками) и протестируйте его(3 балла) **+**

9) Обучите модель, запишите в readme как это предлагается (3 балла) **+**

10) напишите функцию predict, которая примет на вход артефакт/ы от обучения, тестовую выборку(без меток) и запишет предикт, напишите в readme как это сделать (3 балла)  **+**

11) Используется hydra  (https://hydra.cc/docs/intro/) (3 балла - доп баллы) **-**


12) Настроен CI(прогон тестов, линтера) на основе github actions  (3 балла - доп баллы (будем проходить дальше в курсе, но если есть желание поразбираться - welcome) **-**
13) Проведите самооценку, опишите, в какое колво баллов по вашему мнению стоит оценить вашу работу и почему (1 балл доп баллы) **+**
**Итого 29 баллов**

panda1987DS committed:
Версия 1

panda1987DS committed:
Добавила README

panda1987DS committed:
setup and README files

panda1987DS committed:
rewrite README

panda1987DS committed:
rewrite README

goodamigo commented:
Добавьте в PR лейбл hw1

goodamigo commented:
Зачем в predict разбивать на train и test? Данная стадия должна просто выдавать результат по поданным на вход данным. На train и test следует разбивать при прогоне полного пайплайна и train части

goodamigo commented:
В структуре проекта указана папка с ноутбуками, ее не вижу

goodamigo commented on ml_project/src/features/build_features.py:
```diff
@@ -0,0 +1,95 @@
+        return pd.DataFrame(transformer.fit_transform(df))
+
+
+def build_transformer(params: FeatureParams) -> ColumnTransformer:
```
Хотелось бы побольше логов

panda1987DS replied:
Здравствуйте. Добавила побольше логов. Для этого немного переделала логирование, чтобы оно работало через конфигурационный файл

goodamigo commented on ml_project/tests/conftest.py:
```diff
@@ -0,0 +1,105 @@
+
+@pytest.fixture()
+def random_df(categorical_features, numeric_features, sq_features, target_dict) -> pd.DataFrame:
+    return pd.DataFrame({**categorical_features, **numeric_features, **sq_features, **target_dict})
```
хотелось бы как-то похитрее генерить фичи, ближе к тем что реально в датасете про сердце есть

panda1987DS replied:
Не уверена, что получилось как-то хитрее, но расписала все фичи из датасета. Постаралась, чтобы данные получились похожие

panda1987DS committed:
Add more logging
Add logging configuration file
Add notebook
Changed the training pipeline for the prediction mode case

panda1987DS commented:
> Зачем в predict разбивать на train и test? Данная стадия должна просто выдавать результат по поданным на вход данным. На train и test следует разбивать при прогоне полного пайплайна и train части

Исправила.

panda1987DS commented:
> В

Добавила ноутбук

markovivl commented on ml_project/src/train_pipeline.py:
```diff
@@ -0,0 +1,116 @@
+        return model, metrics, cross_val_metric
+    else:
+        return model, None, None
+
```
Очень перегруженная функция, более 50 строк. Хотелось бы подойти чуть более модульно, вынести в отдельные функции.

panda1987DS replied:
Согласна, но поправить не успела)

goodamigo commented on ml_project/tests/test_fit_predict.py:
```diff
@@ -0,0 +1,28 @@
+import models.fit_predict as fit_predict
+
+
+def test_model_train(model_train, training_pipeline_params):
```
Как мне кажется немного странно тренировочную именно часть выносить в конфтест, но не критично

panda1987DS replied:
Глупо получилось. Я думала, модель пригодиться в других тестах.

Спасибо за ревью!!!

goodamigo reviewed (approved):
31 балл (1 доп балл за самооценку)
